Implement retrieval of a size of an index

ewienik opened the pull request:
This is a part of #54.

There is a need for knowledge about a specific index size, how many embeddings are stored in an index. There is a workaround for that functionality which uses a specific table/field. This patch adds HTTP API endpoint to retrieve this information from an index. It introduces an additional index actor API - an index actor is responsible for provided size of an index.

swasik commented on src/httproutes.rs:
```diff
@@ -75,6 +76,36 @@ async fn get_indexes(State(engine): State<Sender<Engine>>) -> response::Json<Vec
 
+#[utoipa::path(
+    get,
+    path = "/api/v1/indexes/{keyspace}/{table}/size",
```
table instead of index

swasik commented on src/httproutes.rs:
```diff
@@ -75,6 +76,36 @@ async fn get_indexes(State(engine): State<Sender<Engine>>) -> response::Json<Vec
+#[utoipa::path(
+    get,
+    path = "/api/v1/indexes/{keyspace}/{table}/size",
+    description = "Get a specific index",
```
index size, not just index

ewienik commented:
Changelog for a38f143:
- fixed commit description
- fixed a `/api/v1/indexes/{keyspace}/{table}/size` endpoint description

<details>
<summary>range diff</summary>

```
1:  46cb453 ! 1:  a38f143 Implement retrieval of a size of an index
    @@ Commit message
         functionality which uses a specific table/field. This patch adds HTTP
         API endpoint to retrieve this information from an index. It introduces
         an additional index actor API - an index actor is responsible for
    -    provided size of an index.
    +    providing size of an index.
     
      ## src/httproutes.rs ##
     @@ src/httproutes.rs: pub(crate) fn new(engine: Sender<Engine>) -> Router {
    @@ src/httproutes.rs: async fn get_indexes(State(engine): State<Sender<Engine>>) ->
      
     +#[utoipa::path(
     +    get,
    -+    path = "/api/v1/indexes/{keyspace}/{table}/size",
    -+    description = "Get a specific index",
    ++    path = "/api/v1/indexes/{keyspace}/{index}/size",
    ++    description = "Get a specific index size",
     +    params(
     +        ("keyspace" = KeyspaceName, Path, description = "A keyspace name for the index"),
     +        ("index" = TableName, Path, description = "An index name")
```

swasik commented on src/httproutes.rs:
```diff
@@ -75,6 +76,36 @@ async fn get_indexes(State(engine): State<Sender<Engine>>) -> response::Json<Vec
 
+#[utoipa::path(
+    get,
+    path = "/api/v1/indexes/{keyspace}/{index}/size",
```
Let us change **size** to **count**. Size is usually measured in bytes (C++ is an exception here). Count clearly suggests the unit (number of elements), it is DB related term (COUNT(*)) and is consistent with Open Search API.

ewienik replied:
fixed in bd0bc17

swasik commented on src/index/usearch.rs:
```diff
@@ -290,3 +295,8 @@ async fn ann(
+
+fn count(idx: Arc<RwLock<usearch::Index>>, tx: oneshot::Sender<CountR>) {
+    tx.send(Ok(idx.read().unwrap().size()))
+        .unwrap_or_else(|_| warn!("index::size: unable to send response"));
```
should we have count here as well?

ewienik replied:
yes, we should. I am changing the warning message

ewienik commented:
Changelog for c664aa6:
- fix warning message
<details>
<summary>range diff</summary>

```
1:  bd0bc17 ! 1:  c664aa6 Implement retrieval of a count of an index
    @@ src/index/usearch.rs: async fn ann(
     +
     +fn count(idx: Arc<RwLock<usearch::Index>>, tx: oneshot::Sender<CountR>) {
     +    tx.send(Ok(idx.read().unwrap().size()))
    -+        .unwrap_or_else(|_| warn!("index::size: unable to send response"));
    ++        .unwrap_or_else(|_| warn!("index::count: unable to send response"));
     +}
     
      ## tests/integration/httpclient.rs ##
```
